In [ ]:
# TensorFlow and tf.keras
import tensorflow as tf
from tensorflow import keras
# Helper libraries
import numpy as np
import matplotlib.pyplot as plt

print(tf.__version__)

2.9.2


In [ ]:
fashion_mnist = tf.keras.datasets.fashion_mnist

(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

4422102/4422102 [==============================] - 0s 0us/step


In [ ]:
class_names = ['T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
               'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot']

In [ ]:
train_images.shape

(60000, 28, 28)

In [ ]:
len(train_labels)

60000

In [ ]:
train_labels

array([9, 0, 0, ..., 3, 0, 5], dtype=uint8)

In [ ]:
test_images.shape

(10000, 28, 28)

In [ ]:
test_images.shape

(10000, 28, 28)

In [ ]:
train_images = train_images / 255.0

test_images = test_images / 255.0

In [ ]:
#Neuronas
#Capa de entrada
model = keras.models.Sequential() #un modelo secuencial
model.add(keras.layers.Flatten(input_shape = (28, 28)))

#Capa oculta
model.add(keras.layers.Dense(128, kernel_initializer = 'he_normal', activation = 'relu')) #1er capa oculta
model.add(keras.layers.Dropout(0.2))

#Capa de Salida
model.add(keras.layers.Dense(10, activation = 'softmax'))

In [ ]:
model.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten_2 (Flatten)         (None, 784)               0         
                                                                 
 dense_2 (Dense)             (None, 128)               100480    
                                                                 
 dropout (Dropout)           (None, 128)               0         
                                                                 
 dense_3 (Dense)             (None, 10)                1290      
                                                                 
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________


In [ ]:
#Compilar el modelo
model.compile(loss = 'sparse_categorical_crossentropy', optimizer = 'sgd', metrics = ['accuracy'])

In [ ]:
#Ajuste del modelo
history = model.fit(train_images, train_labels, epochs = 15, validation_data = (test_images, test_labels), batch_size=1000)

Epoch 1/15
60/60 [==============================] - 2s 21ms/step - loss: 1.9590 - accuracy: 0.3209 - val_loss: 1.5707 - val_accuracy: 0.5737
Epoch 2/15
60/60 [==============================] - 1s 18ms/step - loss: 1.4383 - accuracy: 0.5590 - val_loss: 1.2508 - val_accuracy: 0.6452
Epoch 3/15
60/60 [==============================] - 1s 18ms/step - loss: 1.2079 - accuracy: 0.6175 - val_loss: 1.0806 - val_accuracy: 0.6711
Epoch 4/15
60/60 [==============================] - 1s 17ms/step - loss: 1.0734 - accuracy: 0.6526 - val_loss: 0.9778 - val_accuracy: 0.6892
Epoch 5/15
60/60 [==============================] - 1s 17ms/step - loss: 0.9901 - accuracy: 0.6743 - val_loss: 0.9101 - val_accuracy: 0.7035
Epoch 6/15
60/60 [==============================] - 1s 17ms/step - loss: 0.9321 - accuracy: 0.6892 - val_loss: 0.8617 - val_accuracy: 0.7159
Epoch 7/15
60/60 [==============================] - 1s 17ms/step - loss: 0.8872 - accuracy: 0.7035 - val_loss: 0.8252 - val_accuracy: 0.7260
Epoch 8/15
60

In [ ]:
#Obtener Accuracy
model.evaluate(test_images, test_labels)

313/313 [==============================] - 1s 2ms/step - loss: 0.6832 - accuracy: 0.7690


[0.6831930875778198, 0.7689999938011169]

Para esta primer red, el accuracy es de 0.7690

**Nuestra Neurona**

In [ ]:
#Nuestra propia neurona
#Capa de entrada
model = keras.models.Sequential() #un modelo secuencial
model.add(keras.layers.Flatten(input_shape = (28, 28)))

#Capa oculta 1
model.add(keras.layers.Dense(256, kernel_initializer = 'he_normal', activation = 'relu')) #1er capa oculta
model.add(keras.layers.Dropout(0.2))

#capa oculta 2
model.add(keras.layers.Dense(64, kernel_initializer = 'he_normal', activation = 'relu')) #1er capa oculta
model.add(keras.layers.Dropout(0.2))

#Capa de Salida
model.add(keras.layers.Dense(10, activation = 'softmax'))

In [ ]:
#Compilar el modelo
model.compile(loss = 'sparse_categorical_crossentropy', optimizer = 'adadelta', metrics = ['accuracy'])

In [ ]:
#Ajuste del modelo
history = model.fit(train_images, train_labels, epochs = 30, validation_data = (test_images, test_labels), batch_size=1000)

Epoch 1/30
60/60 [==============================] - 2s 34ms/step - loss: 0.2179 - accuracy: 0.9205 - val_loss: 0.3026 - val_accuracy: 0.8928
Epoch 2/30
60/60 [==============================] - 2s 31ms/step - loss: 0.2163 - accuracy: 0.9206 - val_loss: 0.3019 - val_accuracy: 0.8934
Epoch 3/30
60/60 [==============================] - 2s 30ms/step - loss: 0.2135 - accuracy: 0.9217 - val_loss: 0.3014 - val_accuracy: 0.8940
Epoch 4/30
60/60 [==============================] - 2s 31ms/step - loss: 0.2129 - accuracy: 0.9219 - val_loss: 0.3009 - val_accuracy: 0.8945
Epoch 5/30
60/60 [==============================] - 2s 30ms/step - loss: 0.2130 - accuracy: 0.9207 - val_loss: 0.3005 - val_accuracy: 0.8949
Epoch 6/30
60/60 [==============================] - 2s 30ms/step - loss: 0.2130 - accuracy: 0.9221 - val_loss: 0.3002 - val_accuracy: 0.8953
Epoch 7/30
60/60 [==============================] - 2s 30ms/step - loss: 0.2122 - accuracy: 0.9219 - val_loss: 0.2998 - val_accuracy: 0.8953
Epoch 8/30
60

In [ ]:
#Obtener Accuracy
model.evaluate(test_images, test_labels)

313/313 [==============================] - 1s 3ms/step - loss: 0.2971 - accuracy: 0.8962


[0.2971252501010895, 0.8962000012397766]

En el nuevo modelo la primer capa oculta es de 256 neuronas. Se añade una segunda capa oculta de 64 neuronas con función de activación relu y drop out de 0.2. Al compilar el modelo se usa el optimizador ada delta y se usan 30 epochs.

El accuracy de nuestro modelo es de 0.8962.

El accuracy mejora de 0.7690 a 0.8962

Un incremento de 0.1272 puntos

